### PR TITLE
8256515: javax.xml.XMLEventReader produces incorrect START_DOCUMENT event

### PR DESCRIPTION
**Pull request by MariusVolkhart**
The default implementation of javax.xml.stream.XMLEventReader produced a StartDocument event that always indicated that the "standalone" attribute was set.

The root cause of this was that the com.sun.xml.internal.stream.events.XMLEventAllocatorImpl always set the "standalone" attribute, rather than asking streamReader.standaloneSet() before setting the property of the StartDocumentEvent being created.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256515](https://bugs.openjdk.java.net/browse/JDK-8256515): javax.xml.XMLEventReader produces incorrect START_DOCUMENT event

### Reviewers
 * [Joe Wang](https://openjdk.java.net/census#joehw) (@JoeWang-Java - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1056/head:pull/1056`
`$ git checkout pull/1056`

### CODE REVIEW

File: test/jdk/javax/xml/stream/XmlInputFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.testng.annotations.Test;
+
+import javax.xml.stream.events.StartDocument;
+
+import java.io.StringReader;
+
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class XmlInputFactoryTest {
```
Comment (MariusVolkhart) [inline test/jdk/javax/xml/stream/XmlInputFactoryTest.java:12]: This test is written correctly for regular TestNG, but I don't know if it's "right" for jtreg. I'm happy to make any changes necessary.

**Comment by jerboaa** [comment]
@MariusVolkhart I'll work with you getting the fix in and get the test running using jtreg. Give me a few moments to reproduce.

**Comment by MariusVolkhart** [comment]
Thanks @jerboaa 

I have a JBS number now: [8256515](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8256515). Since opening the PR I've learned that I've gone about the process backwards, and should have found a Sponsor before making the PR. As such, I haven't updated the commit message to avoid sending a premature RFR to the mailing list

**Comment by JoeWang-Java** [comment]
Hi Marius, 

I can't find your name on census, have you signed the OCA?  If you did, pls edit the title to sth. like the following:
8256515: javax.xml.XMLEventReader produces incorrect START_DOCUMENT event

You might also be asked to write /signed in a comment.

Thanks,
Joe

**Comment by jerboaa** [comment]
@MariusVolkhart I think doing a PR is fine before you find a sponsor provided you're fairly certain it's an actual bug. It'll require one to then send an email to the mailing lists asking for a sponsor. That email could potentially reference a PR. That should be fine. The sponsor (committer) will then help you get it integrated. It somewhat depends on the issue. No hard-and-fast rules that I'm aware of in that regard.

**Comment by jerboaa** [comment]
@MariusVolkhart Here is a PR for your branch so the test works with jtreg:
https://github.com/MariusVolkhart/jdk/pull/1

Fails before the JDK patch and passes after.

Run it with:
```
$ rm -rf JTwork JTreport && jtreg -jdk:./build/linux-x86_64-server-release/images/jdk -verbose:summary test/jdk/javax/xml/jaxp/8256515/XmlInputFactoryTest.java
Passed: javax/xml/jaxp/8256515/XmlInputFactoryTest.java
Test results: passed: 1
```
Results are in *.jtr files:

```
$ find JTwork/ -name \*.jtr
JTwork/javax/xml/jaxp/8256515/XmlInputFactoryTest.jtr
```

Or using the test framework of OpenJDK with:

```
$ bash configure --with-jtreg=/path/to/jtreg [...]
$ make test TEST="jdk/javax/xml/jaxp/8256515/XmlInputFactoryTest.java"
```

HTH

**Comment by MariusVolkhart** [comment]
@jerboaa Thanks! That PR was immensely helpful, not just in helping get this change moving, but in helping me understand how to write the tests for next time!

**Comment by JoeWang-Java** [comment]
When you update, pls do a merge from the latest upstream jdk master. The changeset is pretty far behind the master.

File: src/java.xml/share/classes/com/sun/xml/internal/stream/events/XMLEventAllocatorImpl.java [hunk 2]
```diff
@@ -131,7 +131,9 @@ XMLEvent getXMLEvent(XMLStreamReader streamReader) {
                 } else {
                     sdEvent.setDeclaredEncoding(false);
                 }
-                sdEvent.setStandalone(streamReader.isStandalone());
+                if (streamReader.standaloneSet()) {
+                    sdEvent.setStandalone(streamReader.isStandalone());
+                }
```
[the remark this answers is not in the document]
Comment (MariusVolkhart) [reply src/java.xml/share/classes/com/sun/xml/internal/stream/events/XMLEventAllocatorImpl.java]: Updates made. I understood your comment about modifying the `setStandalone` method to mean `StartDocumentEvent#setStandalone`. If it was something else, please let me know!

**Review by JoeWang-Java — APPROVED** [review]

**Comment by JoeWang-Java** [comment]
Hi Marius, 

As the bot suggested, this PR is ready to be integrated. You may issue the integrate command when you're ready, I'll then sponsor it for you.

Thanks for contributing the fix!

Regards,
Joe

**Pull request closed without merging**